core/src/main/java/ai/timefold/solver/core/impl/move/streams/maybeapi/generic/provider/ListChangeMoveProvider.java

```diff
@@ -0,0 +1,118 @@
+                    var valueCount = solutionView.countValues(variableMetaModel, currentPositionInList.entity());
+                    if (valueCount == 1) {
+                        return false; // The value is already in the list, and it is the only one.
+                    } else if (targetPositionInList.index() == valueCount) {
+                        // The value is already in the list, and we are trying to move it past the end of the list.
+                        return false;
+                    } else { // Same list, same position; ignore.
+                        return currentPositionInList.index() != targetPositionInList.index();
+                    }
+                } else { // We can move freely between entities, assuming the target entity accepts the value.
+                    return solutionView.isValueInRange(variableMetaModel, targetPositionInList.entity(), value);
+                }
+            }
+        };
+    }
+
+    @Override
+    public MoveProducer<Solution_> apply(MoveStreamFactory<Solution_> moveStreamFactory) {
+        // For each unassigned value, we need to create a move to assign it to some position of some list variable.
+        // For each assigned value that is not pinned, we need to create:
+        // - A move to unassign it.
+        // - A move to reassign it to another position if assigned.
+        // To assign or reassign a value, we need to create:
+        // - A move for every unpinned value in every entity's list variable to assign the value before that position.
+        // - A move for every entity to assign it to the last position in the list variable.
+        var unpinnedEntities =
+                moveStreamFactory.enumerate(variableMetaModel.entity().type(), variableMetaModel.allowsUnassignedValues());
+        var unpinnedValues = moveStreamFactory.enumerate(variableMetaModel.type(), true)
+                .filter((solutionView, value) -> value == null
+                        || solutionView.getPositionOf(variableMetaModel, value) instanceof PositionInList);
+        var entityValuePairs = unpinnedEntities.join(unpinnedValues, DataJoiners.filtering(isValueInListFilter))
+                .map((solutionView, entity, value) -> {
+                    if (entity == null) { // Null entity means we need to unassign the value.
+                        return ElementPosition.unassigned();
+                    }
+                    var valueCount = solutionView.countValues(variableMetaModel, entity);
+                    if (value == null || valueCount == 0) { // This will trigger assignment of the value at the end of the list.
+                        return ElementPosition.of(entity, valueCount);
+                    } else { // This will trigger assignment of the value immediately before this value.
+                        return solutionView.getPositionOf(variableMetaModel, value);
+                    }
+                })
+                .distinct();
```

Comment:
I wonder if we should always execute a distinct operation, as it may incur additional costs. Producing moves for local search wouldn’t be a problem having duplicate destinations.

Reply:
Yes, it would be a problem. Producing duplicate destinations skews selection probability - some destinations would appear more often than others, and therefore the solver would unfairly prefer some destinations.

Distinct is really not a problem here anyway, as explained in another comment.